### ISSUE TITLE
HTTP API Request/Response Transform Parameter Mapping

### ISSUE THREAD

**Issue by vobarian** [serverless/serverless#9803]
### Use case description

API Gateway v2 HTTP API supports transforming requests and responses (also referred to as parameter mapping). I would like Serverless to provide an easy way to configure this feature. Ideally it could be set on the provider level to apply to all functions and overridden on individual functions.

[Transforming API requests and responses](https://docs.aws.amazon.com/apigateway/latest/developerguide/http-api-parameter-mapping.html)
[AWS::ApiGatewayV2::Integration](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-apigatewayv2-integration.html#cfn-apigatewayv2-integration-requestparameters)

### Proposed solution

Rough draft. Partial configuration, using JS syntax:

```js
const serverlessConfig = {
    provider: {
        httpApi: {
            requestParameterMapping: {
                "append:header.experiment": "$request.path"
            },
            responseParameterMapping: {
                200: {
                    "append:header.cache-control": "no-cache"
                },
                204: {
                    "append:header.cache-control": "no-cache"
                }
            }
        }
    },
    functions: {
        example: {
            httpApi: {
                requestParameterMapping: {
                    "append:header.experiment": "$request.path"
                },
                responseParameterMapping: {
                    200: {
                        "append:header.cache-control": "no-cache"
                    }
                }
            }
        }
    }
}
```

`requestParameterMapping` and `responseParameterMapping` could be named `requestParameters` and `responseParameters` to correspond to the CloudFormation, but I thought the word "mapping" was more explicit and easier to associate with the HTTP API documentation.

I represented both as maps rather than arrays. CloudFormation actually uses a map for `RequestParameters` whereas `ResponseParameters` is an array of objects with separate `Source` and `Destination` properties. Nonetheless, if you specify two `ResponseParameters` with the same `Destination`,  only the second one takes affect. For example, you cannot append two values to the same header; although CloudFormation allows it syntactically, I have determined experimentally that the first mapping is overridden by the second. Representing `responseParameterMapping[status]` as a map in Serverless would avoid unexpected results for users if they tried to specify two mappings for the same destination, since it is not syntactically possible this way. However it could prevent Serverless from supporting this capability if AWS were to change the service someday and start supporting multiple mappings for the same destination.

This example shows how you could specify the mapping on the provider to apply to all functions or set/override it on a single function. If you configure it on both the provider and the function, I'm not sure whether it would be better to merge or completely override the mapping from the provider with that on the function. With merging, if you wanted to remove an inherited mapping key, you would have to set it to null on the function. On the other hand you could make the entire map on the function override the provider, in which case users would have to repeat any entries from the provider they wish to include.

On the function level, the configuration goes directly in `httpApi`, not `events[n].httpApi`, because it is associated with the HTTP API Integration, not the Route.

Response mapping is per HTTP status. It would be nice to have a way to specify the same mapping for multiple statuses. I've shown an example below. The downside to arrays is it becomes possible to specify the same status code more than once and those maps would have to be merged when converting to CloudFormation, and an error raised for duplicate keys.

```js
// Alternative syntax fragment for specifying same mappings for multiple status codes
responseParameterMapping: [
    {
        httpStatus: [ 200, 204 ],
        "append:header.cache-control": "no-cache"
    }
]
```

CloudFormation:

```json
    "HttpApiIntegrationExample": {
      "Type": "AWS::ApiGatewayV2::Integration",
      "Properties": {
        "ApiId": {
          "Ref": "HttpApi"
        },
        "IntegrationType": "AWS_PROXY",
        "IntegrationUri": {
          "Fn::GetAtt": [
            "ExampleLambdaFunction",
            "Arn"
          ]
        },
        "PayloadFormatVersion": "2.0",
        "TimeoutInMillis": 29000,
        "RequestParameters": {
          "append:header.experiment": "$request.path"
        },
        "ResponseParameters": {
          "200": {
            "ResponseParameters": [
              {
                "Destination": "append:header.cache-control",
                "Source": "no-cache"
              }
            ]
          },
          "204": {
            "ResponseParameters": [
              {
                "Destination": "append:header.cache-control",
                "Source": "no-cache"
              }
            ]
          }
        }
      }
    }
```

### Workaround

Until this feature is implemented I work around it using `resources.extensions`:

```js
  resources: {
    extensions: {
      HttpApiIntegrationMain: {
        Properties: {
          ResponseParameters: {
            200: {
              ResponseParameters: [
                {
                  Destination: "append:header.cache-control",
                  Source: "no-cache"
                }
              ]
            },
//...
```

**Comment by pgrzesik** [comment]
Thanks for the proposal @vobarian it looks great overall :+1: 

> Response mapping is per HTTP status. It would be nice to have a way to specify the same mapping for multiple statuses. I've shown an example below. The downside to arrays is it becomes possible to specify the same status code more than once and those maps would have to be merged when converting to CloudFormation, and an error raised for duplicate keys.

While I agree it is nice, I think the syntax is a bit cumbersome and I think we should keep it simpler at first and potentially consider adding support for that if needed. 

> With merging, if you wanted to remove an inherited mapping key, you would have to set it to null on the function. On the other hand you could make the entire map on the function override the provider, in which case users would have to repeat any entries from the provider they wish to include.

I don't have a strong opinion on that - what in your experience is more likely - extending default configuration or overriding it entirely?

**Comment by medikoo** [comment]
@pgrzesik @vobarian if end configuration will turn to be more less 1:1 CloudFormation mapping, maybe we can introduce support for it in more generic way via `cloudFormationExtensions` property (as e.g. proposed here: https://github.com/serverless/serverless/issues/9658#issuecomment-882661172).

That way we'll automatically address any other future extensions

**Comment by vobarian** [comment]
> I don't have a strong opinion on that - what in your experience is more likely - extending default configuration or overriding it entirely?

@pgrzesik I don't have enough experience with this feature to answer that. My intuition is that extending the configuration by merging any mapping on the function with the mapping on the provider, with the function taking precedence, would be a more common use case. I could imagine someone wanting to add a header on all responses from the API while also adding other headers or overriding that header value on one function. For example, one might want to add an HSTS header on all endpoints, so that would go in the provider, but then set different cache-control headers on each function depending on how fresh that data needs to be. It seems less likely that one would *not* want to inherit  the provider mapping. Of course both use cases are supported with either approach, it is just a matter of making the more common one more convenient.

@medikoo While the `cloudFormationExtensions` approach would work, I think it misses out on the benefits of Serverless. Ideally the configuration should be less verbose, more easily discoverable, and generally more user friendly. Otherwise, of course the greatest flexibility can be had using CloudFormation directly without Serverless. I guess it depends on the words "more or less" 1:1. The configuration syntax I propose is simpler than CloudFormation, but it is very similar. So it is a value judgment whether it is worth being its own feature versus a generic extension. Personally I prefer to avoid CloudFormation as it is very verbose. Exposing the feature in Serverless also makes it more discoverable, so users don't have to understand all the CloudFormation resources being used under the hood and look up CloudFormation documentation.

**Comment by pgrzesik** [comment]
Thanks @vobarian - I feel the same way, but I also don't have experience with this feature so I was curious what do you think it's better.

@medikoo that's an interesting suggestion, but as @vobarian mentioned, it would be great if Framework could offer much nicer UX for that configuration piece vs being only the proxy to attach pure CloudFormation properties to resources, in general, I think we should try to avoid pure CloudFormation features as much as possible as they provide steeper learning curve and in most cases are offering worse experience.

**Comment by medikoo** [comment]
> I think we should try to avoid pure CloudFormation features as much as possible as they provide steeper learning curve and in most cases are offering worse experience.

@pgrzesik I understand that point very well, and I honestly shared that for a long time (one year ago I will say same), but I believe that landscape have changed, and we're now in a bit different context:
1. We're stepping here on grounds which we already marked as _legacy_. Our vision is not to invest too much time into this engine but focus more on new one, which will be modular and will bring us multi provider deployments (as described at https://github.com/serverless/serverless/issues/9583)
2. I assume we do not have resources (counting internal + community) to implement a great CF simplification in all areas. Seeing how AWS with CDK evolves, chasing it, is probably not viable for us anymore. This drives me to the point that probably we should focus on most painful CF limitations and quirks (provide a neat, mostly transparent solutions for those), and otherwise on a side let users in all areas to extend CF with CF syntax.

So adding `cloudFormationExtensions` here, is sort of compromise. We leave users with somewhat (not perfect but still a) solution, which doesn't cost us and allows us to move forward with a new solution.

Anyway it's just my train of thought, if community is strongly after CF simplification, we should probably allow it, I would definitely not forcefully block it

**Comment by pgrzesik** [comment]
Thanks for sharing that insight - I'm still more inclined to avoid CloudFormation as much as possible, but I agree with your points and if we have the option to provide support for something in a much faster, while maybe a bit more less convinient way it's definitely something that we should consider, even if it means dealing with more raw CloudFormation. 

>Anyway it's just my train of thought, if community is strongly after CF simplification, we should probably allow it, I would definitely not forcefully block it 

I believe we won't have enough resources to implement that feature internally, so we will rely on community here anyway and I believe both options could be viable (`cloudFormationExtensions` or dedicated "native" syntax), so we could let that up for a decision to a person interested in implementing it - I'm fine with both approaches (though I like "native" approach more in general)

**Comment by marcoscostacoffee** [comment]
Hi, any updates with this feature or a workaround/plugin to have this feature working?

**Comment by marcosfreitas** [comment]
Hi, any updates with this feature or a workaround/plugin to have this feature working?

**Comment by arcaela** [comment]
> ```js
> HttpApiIntegrationMain
> ```

I agree with your proposal, I would like to open a PR with the necessary modifications although I do not know the current status of this feature.
